### PR TITLE
Add logic expressions evaluator function generation

### LINKED ISSUE #13179: Engine - logic expressions evaluator function
| Wazuh version | Component | Install type | Install method   | Platform   |
| ------------- | --------- | ------------ | ---------------- | ---------- |
| 5.0.0         | Engine    | Manager      | Packages/Sources | OS version |

Given a binary expression tree, a function that for each event evaluates the expression must be implemented.

The evaluator should not be recursive, in order to avoid deep stack sizes for complex expressions. A simple algorithm based on [Dijkstra's two stack algorithm](https://hackernoon.com/building-your-own-programming-language-from-scratch-part-ii-dijkstras-two-stack-algorithm) allow to execute all terms/helpers sequentially without recurse. 

The downside is that all terms are evaluated regardless if the expression is already resolved, and trace info is given in suffix notation by default. Thus this could be improved in subsequent iterations.

### PR DESCRIPTION
**Pull request by JavierBejMen**
|Related issue|
|---|
|Closes #13179 |

To support evaluating logic expressions in the engine we need to be able to generate a function that evaluates said expression for each event that comes through. This PR is focused on the functionality needed to generate the function once the expression has been parsed and a tree with each TERM function has been built.
 
Adds evaluator namespace to logicExpression module with types and functions needed to generate the evaluator function based on Dijkstra's two stack algorithm. This functions evaluates a logic expression tree without recurse.

In order to make the API friendlier a function has been added to do the whole process, from parsing, building (with a builder function provided by the programmer to build each TERM function) and generating the evaluator function.

Main changes:
- Added evaluator namespace with all needed functionality and tests
- Updated parser code to be under logicExpression::parser namespace
- Added build function to logicExpression namespace and test
- Added benchmarks for evaluator function

## Tests
image

## Benchmarks
Evaluate `"(PAIR OR ODD AND NOT GREAT5) AND GREAT1"` expression for `N` events.
WIth Release build:
```bash
Run on (16 X 5000 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 0.92, 0.48, 0.43
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_DijkstraEvaluator/1             0.042 us        0.042 us     16977679
BM_DijkstraEvaluator/10            0.043 us        0.043 us     16569865
BM_DijkstraEvaluator/100           0.091 us        0.090 us      7333128
BM_DijkstraEvaluator/1000          0.480 us        0.480 us      1345869
BM_DijkstraEvaluator/10000          4.65 us         4.65 us       138426
BM_DijkstraEvaluator/100000         43.4 us         43.4 us        16090
BM_DijkstraEvaluator/1000000         458 us          458 us         1549
BM_DijkstraEvaluator/10000000       4580 us         4578 us          153
```
With Debug build:
```bash
Run on (16 X 5000 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 0.67, 0.38, 0.39
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
***WARNING*** Library was built as DEBUG. Timings may be affected.
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_DijkstraEvaluator/1              1.13 us         1.13 us       620799
BM_DijkstraEvaluator/10             1.12 us         1.12 us       601762
BM_DijkstraEvaluator/100            1.53 us         1.53 us       457777
BM_DijkstraEvaluator/1000           5.66 us         5.65 us       121702
BM_DijkstraEvaluator/10000          46.1 us         46.0 us        15196
BM_DijkstraEvaluator/100000          454 us          453 us         1542
BM_DijkstraEvaluator/1000000        4571 us         4569 us          152
BM_DijkstraEvaluator/10000000      45872 us        45856 us           15
```

### CODE REVIEW

File: src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include <benchmark/benchmark.h>
+#include <logicExpression/logicExpression.hpp>
+
+using namespace std;
```
Comment (nmkoremblum) [inline src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp:4]:
```suggestion
using std::function;
using std::runtime_error;
using std::string;
```

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+static void BM_DijkstraEvaluator(benchmark::State& state)
+{
+    auto fakeTermBuilder = [](string s) -> function<bool(int)>
+    {
+        if (s == "PAIR")
```
Comment (nmkoremblum) [inline src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp:10]:
```suggestion
        if (s == "EVEN")
```

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        else
+        {
+            throw runtime_error(
+                "Error test fakeBuilder, got unexpected term: " + s);
+        }
+    };
+
+    // Build function
+    auto evaluator = logicExpression::buildDijstraEvaluator<int>(
+        "(PAIR OR ODD AND NOT GREAT5) AND GREAT1", fakeTermBuilder);
```
Comment (nmkoremblum) [inline src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp:47]:
```suggestion
        "(EVEN OR ODD AND NOT GREAT5) AND GREAT1", fakeTermBuilder);
```

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

**Review by JcabreraC — APPROVED** [review]
LGTM !

File: src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp (hunk 1 above)
Comment (JavierBejMen) [reply src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp:4]: Deleted using namespace std and added std:: where needed

File: src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp (hunk 3 above)
Comment (JavierBejMen) [reply src/engine/benchmark/logicExpression/logicExpressionEvaluator_bench.cpp:47]: Fixed

**Review by nmkoremblum — APPROVED** [review]
LGTM

**Pull request merged**